Allow non-root id maps in unprivileged containers.  Taken from
github.com/lxc/lxd/pull/4680.

diff --git a/lxd/container.go b/lxd/container.go
index 77dc2e267..2143136d0 100644
--- a/lxd/container.go
+++ b/lxd/container.go
@@ -243,6 +243,21 @@ func containerValidDeviceConfigKey(t, k string) bool {
 	}
 }

+func allowedUnprivilegedOnlyMap(rawIdmap string) error {
+	rawMaps, err := parseRawIdmap(rawIdmap)
+	if err != nil {
+		return err
+	}
+
+	for _, ent := range rawMaps {
+		if ent.Hostid == 0 {
+			return fmt.Errorf("Cannot map root user into container as LXD was configured to only allow unprivileged containers")
+		}
+	}
+
+	return nil
+}
+
 func containerValidConfig(sysOS *sys.OS, config map[string]string, profile bool, expanded bool) error {
 	if config == nil {
 		return nil
@@ -284,7 +299,10 @@ func containerValidConfig(sysOS *sys.OS, config map[string]string, profile bool,
 	unprivOnly := os.Getenv("LXD_UNPRIVILEGED_ONLY")
 	if shared.IsTrue(unprivOnly) {
 		if config["raw.idmap"] != "" {
-			return fmt.Errorf("raw.idmap can't be set as LXD was configured to only allow unprivileged containers")
+			err := allowedUnprivilegedOnlyMap(config["raw.idmap"])
+			if err != nil {
+				return err
+			}
 		}

 		if shared.IsTrue(config["security.privileged"]) {
